### PR TITLE
Added Header Authentication - Server Part

### PR DESCRIPTION
**Pull request by joewashear007**
This is my attempt at an implementation of header authentication as referenced in issue https://github.com/actualbudget/actual/issues/1092

Looking at how actual passwords work, the front end simply sends a password, so my take was to have my SSO proxy set a `X-ACTUAL-PASSWORD` header that has the value of the my Actual password. To enable this, I add a new env `ACTUAL_LOGIN_METHOD` which should have the value  of "header". 

I updated the the `/account/login` endpoint to check this env and pass along the header value to existing login logic. 

I also added another property to the `/account/needs-bootstrap` that returns the property of the loginMethod. I am working on a second PR (https://github.com/actualbudget/actual/pull/2362) that will adjust the login to automatically the call the `/account/login` endpoint.

Please let me know your thought and recommendations how I can improve and test this. 

Joseph

### CODE REVIEW

**Review by twk3 — CHANGES_REQUESTED** [review]
@joewashear007 can you also look at adding a trusted proxy header auth config option where we can specify cidrs that are allowed to perform header auth? That should help make this a bit more palatable for folks who are security conscious while we still haven't added multi-token support (which would be a first step in allowing us to support better mechanisms like jwts for this).

You should be able to verify the request against the trusted ips using https://github.com/jshttp/proxy-addr#readme which is already available in the project.

**Comment by joewashear007** [comment]
I think that should be do able. I hope to get to this this week,

**Comment by joewashear007** [comment]
I am still wanting to work on this, I just haven't gotten the free time to make it happen.

**Comment by joewashear007** [comment]
@twk3 I was looking at this some more tonight and I want to make I am thinking about this correctly. 

I am assumeing that we want to make a config option that will be for the cidr for auth server. If this is set, when we do an authenticate request, we should check if the sender to see if it is in the range of the cidr we expected, then pass or fail based on that. 

I am not sure how to best use `proxy-addr` for this. If we pass in the config option to the `proxy-addr` then it will spit out the first record of an ip that is not in that list. If that is the case, I am not sure what we do it with it.

I see that `ipaddr` is also included, and the subnetMatch function looks more helpful. But I want to make sure I am not missing something

**Comment by joewashear007** [comment]
@twk3 I think have gotten this part working. I was able to test multiple cases on my local machine and I think it working nicely:
* Password login work
* header login works
* header with invalid password -> redirect to password login
* invalid header or bad proxy trust -> redirect to password login

**Review by twk3 — CHANGES_REQUESTED** [review]
Looking good! I still need to try it out with the client side and review it holistically. In the meantime I've left some minor feedback here.

File: src/load-config.js [hunk 1]
```diff
@@ -85,6 +94,12 @@ if (process.env.NODE_ENV === 'test') {
 
 const finalConfig = {
   ...config,
+  loginMethod: process.env.ACTUAL_LOGIN_METHOD
+    ? process.env.ACTUAL_LOGIN_METHOD.toLowerCase()
+    : config.loginMethod,
+  trustedProxies: process.env.ACTUAL_TRUSTED_PROXIES
+    ? process.env.ACTUAL_TRUSTED_PROXIES.split(',')
```
[the remark this answers is not in the document]
Comment (joewashear007) [reply src/load-config.js:101]: Great idea!

File: upcoming-release-notes/312.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+category: Enhancements
+authors: [joewashear007]
+---
+
+Add option to authenticate with HTTP header from Auth Proxy. Set new env `ACTUAL_LOGIN_METHOD` = 'header' and make sure your proxy set the `X-ACTUAL-PASSWORD` to your actual password. To limit the hosts that are allowed to send the header auth, use the env `ACTUAL_TRUSTED_PROXIES` with a list of command separated Cidrs (default: known internal ip ranges: `[10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, fc00::/7, ::1/128]`).
```
[the remark this answers is not in the document]
Comment (joewashear007) [reply upcoming-release-notes/312.md:6]: Thanks for pointing that out, I will work on a update for documentation site

**Comment by kyrias** [comment]
> can you also look at adding a trusted proxy header auth config option where we can specify cidrs that are allowed to perform header auth? That should help make this a bit more palatable for folks who are security conscious while we still haven't added multi-token support (which would be a first step in allowing us to support better mechanisms like jwts for this).

It seems rather strange to me to do this for this use-case.  Normally you would have whitelisting for trusted proxies when you're actually relying on the proxy to provide correct information, such as if the proxy gives you the name of a pre-authenticated user or to pass in the original client IP as otherwise it might be possible for a malicious client to lie about either of those things, but in this case we're just shifting the exact same information to be sent in a custom header rather than in the `Authorization` header.  There is no proxy-provided information that is assumed to be correct or valid, and any client that can provide one can provide the other.

But for the same reason I'm sort of skeptical over implementing it like this at all, because if we're going to be passing in the same password either way then I don't see any advantage in implementing this instead of having your proxy set the correct `Authorization` header?

**Comment by kyrias** [comment]
Oh sorry, I just realized that I completely forgot how the authentication already works, d'oh.

But well, if we're going to do it this way then I still think that both 1) it would make more sense to use the `Authorization` header than a custom one, and 2) it makes no sense to have this gated by a trusted proxy configuration option, since the value is still authenticated the same way no matter how it's passed in.

**Comment by joewashear007** [comment]
@kyrias thanks for the feedback. I totally agree that would be nice to use the `Authorization` header but when it came to designing a solution I wasn't sure how to make it work without a lot of design/code change. I thought that using the existing login with simply supplying the password through a different channel was a simple low change solution. Do you have any good designs ideal that would make use of the `Authorization` header?

**Comment by kyrias** [comment]
@joewashear007 I'm confused as to what you mean.  What would be the problem with supplying the same password through the `Authorization` header?

**Comment by joewashear007** [comment]
My understanding is that the `Authorization` header was usually Basic Auth (user/password), Bearer token, Oauth token, etc. None of those seemed to quite fix properly and I don't want to overload a standard header that could be used in the future. It standard practice for many API to do authentication with a custom header so I thought it first the best while leaving the path clear for more robust authentication methods in the future

**Comment by joewashear007** [comment]
I guess that is valid. I know Azure DevOps does that with their API, I just thought it was a weird way to do it. My implementation would allow that to be a third option pretty easily.

**Comment by twk3** [comment]
> There is no proxy-provided information that is assumed to be correct or valid, and any client that can provide one can provide the other.

That is a good point. I asked for it to be implemented more as a gating mechanism for controlling how widely header auth is accepted. Not realising/checking that login wasn't already enforcing csrf tokens (so it's not really widening exposure atm). I still think having them it's ideal. But it shouldn't have been required at this stage.

File: src/util/validate-user.js [hunk 3]
```diff
@@ -25,3 +28,26 @@ export default function validateUser(req, res) {
...
   return session;
 }
+
+export function validateAuthHeader(req) {
+  if (config.trustedProxies.length == 0) {
+    return true;
+  }
+
+  let sender = proxyaddr(req, 'uniquelocal');
+  let sender_ip = ipaddr.parse(sender);
```
Comment (twk3) [inline src/util/validate-user.js:38]:
```suggestion
  let sender_ip = ipaddr.process(sender);
```

Caught this during my testing. My own setup apparently was using mapped ips, so process was needed to unmap them.

**Comment by joewashear007** [comment]
Sounds good, I think I might have mixed that up trying to get the TS compile to work.

**Review by twk3 — APPROVED** [review]
Thanks @joewashear007 

Merging the server side now

Would really appreciate it if you could start preparing a documentation PR regarding this for our June release.

**Pull request merged**